### PR TITLE
JSON Escape in JSON log

### PR DESCRIPTION
**Pull request by marcinguy**
My GROK rule had some problems. I realized JSON wasn't properly escaped.

GROK rule:
`
​myParsingrule %{date("yyyy/MM/dd HH:mm:ss"):connect_date} \[%{word:status}\] %{number:id}\#%{number:id2}: \*%{number:value} %{data::json}`

I think only those needs to be escaped.

Examples, how it now works correctly:

```
curl http://127.0.0.1/\"a\?b=\<\>\\

2020/09/03 09:19:50 [error] 25894#0: *3 { "ip":"127.0.0.1","server":"127.0.0.1","uri":"/\"a","vers":"0.56","total_processed":"3","total_blocked":"3","config":"learning","cscore0":"$SQL","score0":"8","cscore1":"$XSS","score1":"24","cscore2":"$TRAVERSAL","score2":"4","zone0":"URL","id0":"1001","var_name0":"","zone1":"ARGS","id1":"1205","var_name1":"b","zone2":"ARGS","id2":"1302","var_name2":"b","zone3":"ARGS","id3":"1303","var_name3":"b" }, client: 127.0.0.1, server: localhost, request: "GET /"a?b=<>\ HTTP/1.1", host: "127.0.0.1"



curl http://127.0.0.1/\"\\a\?b=\<\>\\


2020/09/03 09:19:56 [error] 25894#0: *4 { "ip":"127.0.0.1","server":"127.0.0.1","uri":"/\"\\a","vers":"0.56","total_processed":"4","total_blocked":"4","config":"learning","cscore0":"$SQL","score0":"8","cscore1":"$XSS","score1":"24","cscore2":"$TRAVERSAL","score2":"8","zone0":"URL","id0":"1001","var_name0":"","zone1":"URL","id1":"1205","var_name1":"","zone2":"ARGS","id2":"1205","var_name2":"b","zone3":"ARGS","id3":"1302","var_name3":"b","zone4":"ARGS","id4":"1303","var_name4":"b" }, client: 127.0.0.1, server: localhost, request: "GET /"\a?b=<>\ HTTP/1.1", host: "127.0.0.1"

```

Please review, maybe it can be done better.

The alternative would be to use JSON library, but this would add a dependency, also not sure about the speed.

Thanks,

### CODE REVIEW

**Comment by wargio** [comment]
Hmm. I would say that the encoding is wrong. You should append a `\` for each `\` you find.

**Comment by marcinguy** [comment]
@wargio take a look at example requests. Could be confusing to see. 

First request has zsh escaped `"` and second zsh escaped both `" ` and `\`

Do you see it? 

I append `\` for each `"` and `\`

Or am I missing something?

**Comment by marcinguy** [comment]
@wargio FYI with GROK parsing it worked

You cannot escape the `\` from `"` escaping backslash, if that is what you mean

**Comment by wargio** [comment]
No, what I mean is that you need to backslash each of the \
So `/my/Uri/path/` becomes `\/my\/uri\/` same for any other special chars.

**Comment by marcinguy** [comment]
Ahhh... Seems like you can, but don't have to, per specs:

https://stackoverflow.com/questions/1580647/json-why-are-forward-slashes-escaped#:~:text=JSON%20escapes%20the%20forward%20slash,%2Fb%2Fc%22%7D%20.

@wargio Let me know your decision. I can add it. Will be some code more however on possibly each request. I hope GROK parser will still work :) I need it for GROK parser.

**Comment by marcinguy** [comment]
@wargio Here is JSON RFC:

"characters that MUST be escaped: quotation mark, reverse solidus, and the control characters (U+0000 through U+001F)." 

Source:
https://datatracker.ietf.org/doc/rfc8259/?include_text=1

File: naxsi_src/naxsi_runtime.c [hunk 1]
```diff
@@ -1096,7 +1132,37 @@ ngx_http_output_forbidden_page(ngx_http_request_ctx_t *ctx,
...
+              result = replace_str(result,"\"","\\\"");
+              if(!result)
+              {
+                empty = 1;
+              }
+            }else{
+              empty = 1;  
+            } 
+
+            if(!empty && strlen(json)+strlen(result)<NGX_MAX_ERROR_STR-100-3)
```
Comment (wargio) [inline naxsi_src/naxsi_runtime.c:1154]: what is this magic number ? `NGX_MAX_ERROR_STR-100-3`

**Review by wargio — CHANGES_REQUESTED** [review]

File: naxsi_src/naxsi_runtime.c (hunk 1 above)
Comment (marcinguy) [reply naxsi_src/naxsi_runtime.c:1154]: It is this:
#define MAX_LINE_SIZE (NGX_MAX_ERROR_STR - 100)

and then minus place to terminate JSON (3 bytes, chars)

JSON termination below (3 chars):

https://github.com/nbs-system/naxsi/blob/a8dd4816460314c511f1fd3a322d882dbc2fb9db/naxsi_src/naxsi_runtime.c#L1108

I hope you follow :)

Comment (marcinguy) [reply naxsi_src/naxsi_runtime.c:1170]: cc @wargio

**Comment by marcinguy** [comment]
@wargio Latest changes seem to work correctly. As in first tests:

```
curl http://127.0.0.1/\"a\?b=\<\>\\

2020/09/03 09:19:50 [error] 25894#0: *3 { "ip":"127.0.0.1","server":"127.0.0.1","uri":"/\"a","vers":"0.56","total_processed":"3","total_blocked":"3","config":"learning","cscore0":"$SQL","score0":"8","cscore1":"$XSS","score1":"24","cscore2":"$TRAVERSAL","score2":"4","zone0":"URL","id0":"1001","var_name0":"","zone1":"ARGS","id1":"1205","var_name1":"b","zone2":"ARGS","id2":"1302","var_name2":"b","zone3":"ARGS","id3":"1303","var_name3":"b" }, client: 127.0.0.1, server: localhost, request: "GET /"a?b=<>\ HTTP/1.1", host: "127.0.0.1"



curl http://127.0.0.1/\"\\a\?b=\<\>\\


2020/09/03 09:19:56 [error] 25894#0: *4 { "ip":"127.0.0.1","server":"127.0.0.1","uri":"/\"\\a","vers":"0.56","total_processed":"4","total_blocked":"4","config":"learning","cscore0":"$SQL","score0":"8","cscore1":"$XSS","score1":"24","cscore2":"$TRAVERSAL","score2":"8","zone0":"URL","id0":"1001","var_name0":"","zone1":"URL","id1":"1205","var_name1":"","zone2":"ARGS","id2":"1205","var_name2":"b","zone3":"ARGS","id3":"1302","var_name3":"b","zone4":"ARGS","id4":"1303","var_name4":"b" }, client: 127.0.0.1, server: localhost, request: "GET /"\a?b=<>\ HTTP/1.1", host: "127.0.0.1"
```

Should me more optimal, only to call unescaping when needed for `\` and/or for `"`

**Comment by marcinguy** [comment]
@wargio Added tests. But somehow the test plan complains on the amount of tests. Do you know how to fix it?

**Comment by marcinguy** [comment]
@wargio Fixed test. Tests are green now also

File: naxsi_src/naxsi_runtime.c [hunk 2]
```diff
@@ -1007,6 +1007,42 @@ ngx_int_t ngx_http_nx_log(ngx_http_request_ctx_t *ctx,
...
+  for (i = 0; s[i] != '\0'; i++) { 
+    if (strstr(&s[i], oldW) == &s[i]) { 
+      cnt++; 
+      // Jumping to index after the old word. 
+      i += oldWlen - 1; 
+    } 
+  } 
+
+  // Making new string of enough length 
+  result = (char*)malloc(i + cnt * (newWlen - oldWlen) + 1);
```
Comment (wargio) [inline naxsi_src/naxsi_runtime.c:1029]: check if pointer is not null.

File: naxsi_src/naxsi_runtime.c [hunk 3]
```diff
@@ -1096,7 +1132,60 @@ ngx_http_output_forbidden_page(ngx_http_request_ctx_t *ctx,
...
+          int escaped = 0;
+
+          if(strstr(value,"\\"))
+          { 
+            //escaping JSON
+            result = replace_str(value,"\\","\\\\");
+            if(!result)
+              empty = 1;  
+
+            if(empty || strlen(json)+strlen(result)>NGX_MAX_ERROR_STR-100-3)
```
Comment (wargio) [inline naxsi_src/naxsi_runtime.c:1148]: Please use a macro and add some parenthesis like:
`if (empty || ((strlen(json) + strlen(result)) > (NGX_MAX_ERROR_STR-100-3)) )`

File: naxsi_src/naxsi_runtime.c (hunk 1 above)
Comment (wargio) [reply naxsi_src/naxsi_runtime.c:1172]: do as i mentioned above

**Review by wargio — CHANGES_REQUESTED** [review]
looks really good. it just needs a few edits.

File: naxsi_src/naxsi_runtime.c (hunk 3 above)
Comment (marcinguy) [reply naxsi_src/naxsi_runtime.c:1148]: @wargio do you mean like: 

#define max_json(json,result) ((strlen(json) + strlen(result)) > (NGX_MAX_ERROR_STR-100-3)))

?

File: naxsi_src/naxsi_runtime.c (hunk 2 above)
Comment (marcinguy) [reply naxsi_src/naxsi_runtime.c:1029]: OK

File: naxsi_src/naxsi_runtime.c (hunk 3 above)
Comment (wargio) [reply naxsi_src/naxsi_runtime.c:1148]: yes, that would be lovely.

Comment (marcinguy) [reply naxsi_src/naxsi_runtime.c]: OK

**Comment by marcinguy** [comment]
@wargio Changed per your requests. Tests seems green.

**Review by marcinguy — COMMENTED** [review]
Done

**Review by wargio — APPROVED** [review]

**Pull request merged**